eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/ServiceClientInstantiationCheck.java

```diff
@@ -0,0 +1,220 @@
+     *
+     * @param classDefToken the CLASS_DEF AST node
+     * @return true if the class is annotated with @ServiceClient, false otherwise.
+     */
+    private boolean hasServiceClientAnnotation(DetailAST classDefToken) {
+        // Always has MODIFIERS node
+        final DetailAST modifiersToken = classDefToken.findFirstToken(TokenTypes.MODIFIERS);
+
+        for (DetailAST ast = modifiersToken.getFirstChild(); ast != null; ast = ast.getNextSibling()) {
+            if (ast.getType() != TokenTypes.ANNOTATION) {
+                continue;
+            }
+            // One class could have multiple annotations, return true if found one.
+            final DetailAST annotationIdent = ast.findFirstToken(TokenTypes.IDENT);
+            if (annotationIdent != null && SERVICE_CLIENT.equals(annotationIdent.getText())) {
+                isAsync = isAsyncServiceClient(ast);
+                return true;
+            }
+        }
+        // If no @ServiceClient annotated with this class, return false
+        return false;
+    }
+
+    /**
+     *  Checks for public or protected constructor for the service client class.
+     *  Log error if the service client has public or protected constructor.
+     *
+     * @param ctorToken the CTOR_DEF AST node
+     */
+    private void checkConstructor(DetailAST ctorToken) {
+        final DetailAST modifiersToken = ctorToken.findFirstToken(TokenTypes.MODIFIERS);
+        // find constructor's modifier accessibility, no public or protected constructor
+        final AccessModifier accessModifier = CheckUtil.getAccessModifierFromModifiersToken(modifiersToken);
+        if (accessModifier.equals(AccessModifier.PUBLIC) || accessModifier.equals(AccessModifier.PROTECTED)) {
+            log(modifiersToken, "@ServiceClient class should not have any public or protected constructor.");
+        }
+    }
+
+    /**
+     * Checks for public static method named 'builder'. Should avoid to use method name, 'builder'.
+     *
+     * @param methodDefToken the METHOD_DEF AST node
+     */
+    private void checkMethodName(DetailAST methodDefToken) {
+        final DetailAST methodNameToken = methodDefToken.findFirstToken(TokenTypes.IDENT);
+        if (!BUILDER.equals(methodNameToken.getText())) {
+            return;
+        }
+
+        final DetailAST modifiersToken = methodDefToken.findFirstToken(TokenTypes.MODIFIERS);
+        // find method's modifier accessibility, should not have a public static method called 'builder'
+        final AccessModifier accessModifier = CheckUtil.getAccessModifierFromModifiersToken(modifiersToken);
+        if (accessModifier.equals(AccessModifier.PUBLIC) && modifiersToken.branchContains(TokenTypes.LITERAL_STATIC)) {
+            log(modifiersToken, "@ServiceClient class should not have a public static method named ''builder''.");
+        }
+    }
+
+    /**
+     * Checks for the variable field of the subclass of ServiceClient.
```

Comment:
Remove the two lines and replace with: "Checks that the field variables in the subclass of ServiceClient are final. ServiceClients should be immutable."